### PR TITLE
ISSUE #4784 - Custom tickets: Readonly for UI & unique property in templates

### LINKED ISSUE #4784: Custom tickets - Readonly for UI & unique property in templates
### Description
For supporting new features (#4777) we need to add `readOnlyOnUI` and `unique` configuration for properties in custom tickets. 
- `readOnlyOnUI` from the frontend standpoint is identical to `readOnly`. Both configs applies.
- `unique` will add a validation to ensure that the value from that property is not repeated throughout all the tickets of that type that belongs to the container (or federation).

### Goals
- [ ] Users see `readOnlyOnUI` properties as readOnly
- [ ] Users can interact with `unique` properties and have them validated 

### Tasks
- [ ] In template selector replace `readOnly` with the value `readOnlyOnUI || readOnly`
- [ ] In the ticket form add a validation for testing the uniqueness of `unique` properties and provide an error message when that happens.

### Template schema
```` js
{
  properties: [
    // This property will be read only on the UI, but editable via the API
    {
        name: "my property",
        type: "text",
        readOnlyOnUI: true
    }
    // This property will be validated against existing tickets of the same template and ensure the value is unique
    // If value is undefined, the check will not happen.
    {
        name: "my property",
        type: "text",
        unique: true
    }
    // This property wll not be editable on the UI, and it has to be unique.
    {
        name: "my property",
        type: "text",
        unique: true,
        readOnlyOnUI: true
    }
    
  ]
//other ticket template values
}

````

### PR DESCRIPTION
**Pull request by The-Daniel**
This fixes #4784

#### Description
Custom tickets properties can now be 'readOnlyOnUI' and 'unique'. This PR addresses the frontend for this change.

#### Test cases

Check that both new properties work

#### Notes
<details><summary>Template with unique and readOnlyUI properties</summary>
<p>

```
{
    "_id":"0f624abe-5313-403e-9723-f555cd961b03",
    "modules":[
        {
            "properties":[
                {
                    "type":"boolean",
                    "default":false,
                    "immutable":true,
                    "name":"Boolean"
                },
                {
                    "type":"text",
                    "immutable":true,
                    "name":"Text"
                },
                {
                    "type":"longText",
                    "immutable":true,
                    "name":"Long Text"
                },
                {
                    "type":"date",
                    "immutable":true,
                    "name":"Date"
                },
                {
                    "type":"number",
                    "immutable":true,
                    "name":"Number"
                },
                {
                    "type":"image",
                    "immutable":true,
                    "name":"Image"
                },
                {
                    "type":"coords",
                    "immutable":true,
                    "name":"Coordinates"
                },
                {
                    "type":"oneOf",
                    "immutable":true,
                    "values":"jobsAndUsers",
                    "name":"One of"
                },
                {
                    "type":"manyOf",
                    "immutable":true,
                    "values":"riskCategories",
                    "name":"Many of"
                },
                {
                    "name":"Text (Prefilled)",
                    "type":"text",
                    "deprecated":true,
                    "default":"This text input has a default value",
                    "immutable":true
                }
            ],
            "name":"Immutable properties"
        },
        {
            "type":"safetibase",
            "properties":[
                {
                    "type":"boolean",
                    "default":false,
                    "name":"Risk Assessed"
                }
            ]
        },
        {
            "properties":[
                {
                    "type":"coords",
                    "name":"Colour Undefined"
                },
                {
                    "type":"coords",
                    "color":[
                        255,
                        0,
                        0
                    ],
                    "name":"Colour Specified (red)"
                },
                {
                    "type":"boolean",
                    "default":false,
                    "name":"Boolean"
                },
                {
                    "type":"coords",
                    "name":"Colour By Property (Switch)",
                    "color":{
                        "property":{
                            "module":"Coloured Pins",
                            "name":"Boolean"
                        },
                        "mapping":[
                            {
                                "default":[
                                    245,
                                    245,
                                    220
                                ]
                            },
                            {
                                "value":true,
                                "color":[
                                    160,
                                    255,
                                    0
                                ]
                            }
                        ]
                    }
                },
                {
                    "type":"coords",
                    "color":{
                        "property":{
                            "name":"Status"
                        },
                        "mapping":[
                            {
                                "default":[
                                    188,
                                    190,
                                    202
                                ]
                            },
                            {
                                "value":"Open",
                                "color":[
                                    107,
                                    119,
                                    140
                                ]
                            },
                            {
                                "value":"In Progress",
                                "color":[
                                    113,
                                    86,
                                    255
                                ]
                            },
                            {
                                "value":"For Approval",
                                "color":[
                                    2,
                                    136,
                                    209
                                ]
                            },
                            {
                                "value":"Closed",
                                "color":[
                                    46,
                                    125,
                                    50
                                ]
                            },
                            {
                                "value":"Void",
                                "color":[
                                    0,
                                    0,
                                    0
                                ]
                            }
                        ]
                    },
                    "name":"Colour By Property (Status)"
                },
                {
                    "type":"text",
                    "name":"Colour Textbox"
                },
                {
                    "type":"coords",
                    "color":{
                        "property":{
                            "module":"Coloured Pins",
                            "name":"Colour Textbox"
                        },
                        "mapping":[
                            {
                                "default":[
                                    255,
                                    255,
                                    255
                                ]
                            },
                            {
                                "value":"black",
                                "color":[
                                    0,
                                    0,
                                    0
                                ]
                            },
                            {
                                "value":"red",
                                "color":[
                                    255,
                                    0,
                                    0
                                ]
                            },
                            {
                                "value":"green",
                                "color":[
                                    0,
                                    255,
                                    0
                                ]
                            },
                            {
                                "value":"blue",
                                "color":[
                                    0,
                                    0,
                                    255
                                ]
                            },
                            {
                                "value":"hotpink",
                                "color":[
                                    255,
                                    105,
                                    180
                                ]
                            }
                        ]
                    },
                    "name":"Colour By Property (Textbox)"
                },
                {
                    "name":"Colour Threshold Test",
                    "type":"oneOf",
                    "values":[
                        "PRIMARY_CONTRAST",
                        "TERTIARY_LIGHTEST",
                        "TERTIARY_LIGHTER",
                        "BASE_LIGHTEST",
                        "BASE_LIGHTER",
                        "BASE_LIGHT",
                        "ERROR_LIGHTEST",
                        "BLACK"
                    ]
                },
                {
                    "name":"Colour by Property (Select)",
                    "type":"coords",
                    "color":{
                        "property":{
                            "module":"Coloured Pins",
                            "name":"Colour Threshold Test"
                        },
                        "mapping":[
                            {
                                "default":[
                                    250,
                                    229,
                                    154
                                ]
                            },
                            {
                                "value":"PRIMARY_CONTRAST",
                                "color":[
                                    255,
                                    255,
                                    255
                                ]
                            },
                            {
                                "value":"TERTIARY_LIGHTEST",
                                "color":[
                                    242,
                                    246,
                                    252
                                ]
                            },
                            {
                                "value":"TERTIARY_LIGHTER",
                                "color":[
                                    232,
                                    238,
                                    248
                                ]
                            },
                            {
                                "value":"BASE_LIGHTEST",
                                "color":[
                                    208,
                                    217,
                                    235
                                ]
                            },
                            {
                                "value":"BASE_LIGHTER",
                                "color":[
                                    193,
                                    200,
                                    213
                                ]
                            },
                            {
                                "value":"BASE_LIGHT",
                                "color":[
                                    188,
                                    190,
                                    202
                                ]
                            },
                            {
                                "value":"ERROR_LIGHTEST",
                                "color":[
                                    249,
                                    236,
                                    236
                                ]
                            },
                            {
                                "value":"BLACK",
                                "color":[
                                    0,
                                    0,
                                    0
                                ]
                            }
                        ]
                    }
                }
            ],
            "name":"Coloured Pins"
        },
        {
            "name":"Unique",
            "properties":[
                {
                    "type":"oneOf",
                    "values":"jobsAndUsers",
                    "name":"One of",
                    "unique":true
                },
                {
                    "type":"text",
                    "name":"Text",
                    "unique":true
                },
                {
                    "type":"number",
                    "name":"Number",
                    "unique":true
                }
            ]
        },
        {
            "name":"immutable",
            "deprecated":true,
            "properties":[
                {
                    "type":"boolean",
                    "default":false,
                    "name":"Boolean"
                },
                {
                    "type":"text",
                    "name":"Text"
                },
                {
                    "type":"text",
                    "default":"This text input has a default value",
                    "name":"Text (Prefilled)"
                },
                {
                    "type":"longText",
                    "name":"Long Text"
                },
                {
                    "type":"date",
                    "name":"Date"
                },
                {
                    "type":"number",
                    "name":"Number"
                },
                {
                    "type":"image",
                    "name":"Image"
                },
                {
                    "type":"coords",
                    "name":"Coordinates"
                },
                {
                    "type":"oneOf",
                    "values":"jobsAndUsers",
                    "name":"One of"
                },
                {
                    "type":"manyOf",
                    "values":"riskCategories",
                    "name":"Many of"
                }
            ]
        }
    ],
    "properties":[
        {
            "name":"Colour by Property (Priority)",
            "type":"coords",
            "color":{
                "property":{
                    "name":"Priority"
                },
                "mapping":[
                    {
                        "default":[
                            107,
                            119,
                            140
                        ]
                    },
                    {
                        "value":"Low",
                        "color":[
                            2,
                            136,
                            209
                        ]
                    },
                    {
                        "value":"Medium",
                        "color":[
                            237,
                            108,
                            2
                        ]
                    },
                    {
                        "value":"High",
                        "color":[
                            190,
                            67,
                            67
                        ]
                    }
                ]
            }
        },
        {
            "type":"text",
            "name":"readOnlyOnUI",
            "readOnlyOnUI":true
        }
    ],
    "config":{
        "comments":true,
        "issueProperties":true,
        "pin":{
            "color":{
                "property":{
                    "module":"safetibase",
                    "name":"Risk Assessed"
                },
                "mapping":[
                    {
                        "default":[
                            235,
                            0,
                            0
                        ]
                    },
                    {
                        "value":true,
                        "color":[
                            0,
                            235,
                            0
                        ]
                    }
                ]
            }
        },
        "status":{
            "values":[
                {
                    "name":"Awaiting triage",
                    "type":"open"
                },
                {
                    "name":"Triaged (This option is extra long to test for overflow)",
                    "type":"open"
                },
                {
                    "name":"Active",
                    "type":"active"
                },
                {
                    "name":"To verify",
                    "type":"review"
                },
                {
                    "name":"Approved",
                    "type":"done"
                },
                {
                    "name":"Won't fix",
                    "type":"void"
                }
            ],
            "default":"Awaiting triage"
        }
    },
    "code":"DMO",
    "name":"Demo Template"
}
```
</p>
</details>

### CODE REVIEW

File: frontend/src/v5/store/tickets/tickets.types.ts [hunk 1]
```diff
@@ -24,7 +24,9 @@ export interface PropertyDefinition {
 	type: PropertyTypeDefinition;
 	values?: string[] | 'jobsAndUsers' | 'riskCategories';
 	default?: string;
+	unique?: boolean;
```
Comment (sanmont3drepo) [inline frontend/src/v5/store/tickets/tickets.types.ts:27]: When there's a unique error triggered, after closing the modal, the save ticket button stays in processing state.
cant edit

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

**Review by sanmont3drepo — APPROVED** [review]

**Review by sanmont3drepo — APPROVED** [review]

**Review by sanmont3drepo — APPROVED** [review]

File: frontend/src/v5/store/tickets/tickets.types.ts [hunk 2]
```diff
@@ -25,7 +25,9 @@ export interface PropertyDefinition {
...
 	readOnly?: boolean;
+	readOnlyOnUI?: boolean;
```
Comment (Amantini1997) [inline frontend/src/v5/store/tickets/tickets.types.ts:30]: Imagine a ticket with 2 properties: A (which is unique) and B.
If I try to update A to a value that already exists elsewhere, I get an error message. But if then I update B, should A be reverted to its previous value, so that B is updated? (At least, this what normally happens with error, such as with a required field).

Also, note that if I update A to a non-unique value and I try to update B, I get 2 error modals


Uploading 2024-03-13 17-58-34.mp4…

**Review by Amantini1997 — CHANGES_REQUESTED** [review]

File: frontend/src/v5/store/tickets/tickets.sagas.ts [hunk 3]
```diff
@@ -166,7 +166,7 @@ export function* updateTicket({ teamspace, projectId, modelId, ticketId, ticket,
 	}
 }
 
-export function* createTicket({ teamspace, projectId, modelId, ticket, isFederation, onSuccess }: CreateTicketAction) {
+export function* createTicket({ teamspace, projectId, modelId, ticket, isFederation, onSuccess, onError }: CreateTicketAction) {
```
Comment (sanmont3drepo) [inline frontend/src/v5/store/tickets/tickets.sagas.ts:170]: - When trying to change a unique property it sends the request twice
- It also tries sending it each time I change focus in the form 
triple sent

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/store/tickets/tickets.helpers.ts [hunk 4]
```diff
@@ -287,3 +287,41 @@ export const fillOverridesIfEmpty = (values: Partial<ITicket>) => {
...
+				someArray: ['a', 'b', 'c'],
+			}
+		}
+		properties: {
+			foo: true,
+		}
+	})
+	returns ['modules.safetibase.location', 'modules.safetibase.someArray', 'properties.foo']
+*/
+export const getAllPaths = (root) => {
```
Comment (sanmont3drepo) [inline frontend/src/v5/store/tickets/tickets.helpers.ts:305]: I don't think this is being used.

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

**Review by sanmont3drepo — APPROVED** [review]

**Review by Amantini1997 — APPROVED** [review]
